### PR TITLE
Update CreateProjectInfo to accept relative analyzer paths

### LINKED ISSUE #20187: CommandLineProject.CreateProjectInfo cannot handle relative analyzer references
I try to use CommandLineProject.CreateProjectInfo to create a ProjectInfo from a command line that contains an analyzer reference:
```
/analyzer:..\\packages\\Microsoft.CodeAnalysis.Analyzers.1.1.0\\analyzers\\dotnet\\cs\\Microsoft.CodeAnalysis.Analyzers.dll
```

Roslyn 1.3.4 successfully handles such references, however  Roslyn 2.2.0  does not convert a relative path 
to an absolute path for analyzer references. Thus creation of ProjectInfo fails with an exception:
```
System.ArgumentException occurred
  HResult=0x80070057
  Message=Absolute path expected.
  Source=Microsoft.CodeAnalysis.Workspaces
  StackTrace:
   at Roslyn.Utilities.CompilerPathUtilities.RequireAbsolutePath(String path, String argumentName) in G:\csharp\roslyn\src\Workspaces\Core\Portable\Utilities\CompilerUtilities\CompilerPathUtilities.cs:line 19
   at Microsoft.CodeAnalysis.AnalyzerAssemblyLoader.AddDependencyLocation(String fullPath) in G:\csharp\roslyn\src\Compilers\Core\Portable\DiagnosticAnalyzer\AnalyzerAssemblyLoader.cs:line 30
   at Microsoft.CodeAnalysis.CommandLineProject.CreateProjectInfo(String projectName, String language, IEnumerable`1 commandLineArgs, String projectDirectory, Workspace workspace) in G:\csharp\roslyn\src\Workspaces\Core\Desktop\Workspace\CommandLineProject.cs:line 57
   at Microsoft.CodeAnalysis.CommandLineProject.CreateProjectInfo(String projectName, String language, String commandLine, String baseDirectory, Workspace workspace) in G:\csharp\roslyn\src\Workspaces\Core\Desktop\Workspace\CommandLineProject.cs:line 179
   at OpenCmdLineTestNuget.Program.Main(String[] args) in C:\Users\vedun\Documents\Visual Studio 
```

### PR DESCRIPTION
**Pull request by vedun-z**
I try to use CommandLineProject.CreateProjectInfo to create a ProjectInfo from a command line that contains an analyzer reference:
```
/analyzer:..\\packages\\Microsoft.CodeAnalysis.Analyzers.1.1.0\\analyzers\\dotnet\\cs\\Microsoft.CodeAnalysis.Analyzers.dll
```

Roslyn 1.3.4 successfully handles such references, however  Roslyn 2.2.0  does not convert a relative path 
to an absolute path for analyzer references. Thus creation of ProjectInfo fails with an exception:
```
System.ArgumentException occurred
  HResult=0x80070057
  Message=Absolute path expected.
  Source=Microsoft.CodeAnalysis.Workspaces
  StackTrace:
   at Roslyn.Utilities.CompilerPathUtilities.RequireAbsolutePath(String path, String argumentName) in G:\csharp\roslyn\src\Workspaces\Core\Portable\Utilities\CompilerUtilities\CompilerPathUtilities.cs:line 19
   at Microsoft.CodeAnalysis.AnalyzerAssemblyLoader.AddDependencyLocation(String fullPath) in G:\csharp\roslyn\src\Compilers\Core\Portable\DiagnosticAnalyzer\AnalyzerAssemblyLoader.cs:line 30
   at Microsoft.CodeAnalysis.CommandLineProject.CreateProjectInfo(String projectName, String language, IEnumerable`1 commandLineArgs, String projectDirectory, Workspace workspace) in G:\csharp\roslyn\src\Workspaces\Core\Desktop\Workspace\CommandLineProject.cs:line 57
   at Microsoft.CodeAnalysis.CommandLineProject.CreateProjectInfo(String projectName, String language, String commandLine, String baseDirectory, Workspace workspace) in G:\csharp\roslyn\src\Workspaces\Core\Desktop\Workspace\CommandLineProject.cs:line 179
   at OpenCmdLineTestNuget.Program.Main(String[] args) in C:\Users\vedun\Documents\Visual Studio 
```

To resolve the issue I have changed
```
 analyzerLoader.AddDependencyLocation(path) // path to an analyzer reference
```
to
```
analyzerLoader.AddDependencyLocation(relativePathResolver.ResolvePath(path, null))
```

**Bugs this fixes:**

Fixes #20187

**Risk**

The patch affects only code that does not work properly anyway.

**Performance impact**

No performance impact

**Is this a regression from a previous update?**

Yes, 1.3.4 -> 2.2.0 regression.
**Root cause analysis:**

Roslyn does not have a test that checks creation of ProjectInfo from a command line with analyzer references.

**How was the bug found?**

Migration from 1.3.4 to 2.2.0 breaks our internal tests.

### CODE REVIEW

**Comment by Pilchie** [comment]
@vedun-z thanks for the contribution!

@dotnet/roslyn-analysis and @dotnet/roslyn-ide for review

**Comment by tmeschter** [comment]
This looks right to me, but I'd like to understand how we regressed this.

**Comment by vedun-z** [comment]
@tmeschter 
In Roslyn 1.3.2 implementations of  IAnalyzerAssemblyLoader.AddDependencyLocation() do not check that the fullPath is actually absolute, they just add a reference to the dictionary:

https://github.com/dotnet/roslyn/blob/version-1.3.2/src/Compilers/Shared/AbstractAnalyzerAssemblyLoader.cs#L44

Probably analyzer references do not work in 1.3.2 too, but since there are no exceptions and we don't care about analyzer references in our project, we did not notice that.

File: src/Workspaces/Core/Desktop/Workspace/CommandLineProject.cs [hunk 1]
```diff
@@ -52,9 +54,10 @@ public static ProjectInfo CreateProjectInfo(string projectName, string language,
             }
 
             // resolve all analyzer references.
+
             foreach (var path in commandLineArguments.AnalyzerReferences.Select(r => r.FilePath))
             {
-                analyzerLoader.AddDependencyLocation(path);
+                analyzerLoader.AddDependencyLocation(relativePathResolver.ResolvePath(path, null));
```
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/CommandLineProject.cs]: can you make null to named argument? basePath: null

**Comment by heejaechang** [comment]
👍

**Comment by heejaechang** [comment]
retest windows_eta_open_prtest please

File: src/Workspaces/Core/Desktop/Workspace/CommandLineProject.cs (hunk 1 above)
Comment (vedun-z) [inline src/Workspaces/Core/Desktop/Workspace/CommandLineProject.cs]: done

**Comment by Pilchie** [comment]
This is actually not where I've been working (I've been up in the VS layer lately), but we do have a [`CommandLineProjectTests`](http://source.roslyn.io/#Roslyn.Services.UnitTests/WorkspaceTests/CommandLineProjectTests.cs,dd41942d99b809da)...

**Comment by vedun-z** [comment]
@Pilchie @jasonmalinowski If you don't mind I'm going to add a test to CommandLineProjectTests.

**Comment by vedun-z** [comment]
@Pilchie @jasonmalinowski
I have an issue with creating the test. RelativePathResolver returns null if a path does not exist. Tests in CommandLineProjectTests do not create files, they just use paths to files that do not exist.

**Comment by Pilchie** [comment]
You can inherity from [`TestBase`](http://source.roslyn.io/#Roslyn.Test.Utilities/TestBase.cs,772742e87ab6ab9d,references), and it has ways to create real temp files that will get cleaned up when it is disposed. (Or see how it uses `TempFile` and `TempDirectory`.

**Comment by vedun-z** [comment]
I don't mind about file content so I use `typeof(object).Assembly.Location`. Please, check my new commit.

**Comment by vedun-z** [comment]
@jasonmalinowski @Pilchie  @heejaechang Almost a week has passed since my last commit. Can you guys have a look at my test?

File: src/Workspaces/CoreTest/WorkspaceTests/CommandLineProjectTests.cs [hunk 2]
```diff
@@ -110,5 +111,32 @@ public void TestAdditionalFiles()
...
+            {
+                var assemblyBaseBaseDir = PathUtilities.GetDirectoryName(assemblyBaseDir);
+                assemblyBaseDirName = PathUtilities.GetRelativePath(assemblyBaseBaseDir, assemblyBaseDir);
+            }
+            else // e.g. pathToAssembly = C:\mscorlib.dll.
+            {
+                assemblyBaseDirName = "..";
+            }
+            var relativePath = PathUtilities.CombinePossiblyRelativeAndRelativePaths(
+                PathUtilities.CombinePossiblyRelativeAndRelativePaths("..", assemblyBaseDirName), assemblyName);
```
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/WorkspaceTests/CommandLineProjectTests.cs]: Is it necessary for this test to actually create a path that has a prefix like this? Is:

    var pathToAssembly = typeof(object).Assembly.Location;
    var assemblyBaseDir = Path.GetDirectoryName(pathToAssembly);
    string commandLine = @"foo.cs /a:" + Path.GetFileName(pathToAssembly);

not sufficient?

It just seems like a lot of non-trivial setup that is really easy to get lost in really quickly.

**Comment by jasonmalinowski** [comment]
@vedun-z Thanks for the ping! Did have one remaining question.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest this please.

**Comment by vedun-z** [comment]
@jasonmalinowski  I outwitted myself with relative path creation :-D.  I hope that using System.IO.Path is ok. Just commited simplified version.

**Comment by jasonmalinowski** [comment]
I'm sure @tmat will come by and tell me why I shouldn't be using `Path`, but it's a unit test so we're fine here. And yeah, that logic I can actually follow now. :smile:

**Comment by jasonmalinowski** [comment]
Can we get a second review from @Pilchie or @dotnet/roslyn-ide before we merge this?

**Comment by sharwell** [comment]
@jasonmalinowski I'd love to see a comment indicating when the regression was introduced, should someone have that information. Otherwise, happy to see a new test added and 👍

**Comment by jasonmalinowski** [comment]
@sharwell It probably never worked. :smile:

**Comment by sharwell** [comment]
@jasonmalinowski The ask mode template says it worked in 1.3.4.

**Comment by jasonmalinowski** [comment]
Thanks @vedun-z for the contribution!

**Comment by vedun-z** [comment]
@jasonmalinowski, you are welcome!

**Pull request merged**